src/filter.ts

```diff
@@ -387,41 +387,47 @@ type NotUndefined = string | number | boolean | SomeObject;
  * Given a FilterQuery, returns an object that, when intersected with an Update,
  * marks those properties as required that are guaranteed to exist.
  */
-type RunQuery<Q extends string> = L1Combinations<Q, L1Parts<Q>>;
+type RunQuery<Q extends string> = L1Discriminator<Q, L1Parts<Q>>;
+
+// gets all L1 query snippets
+type L1Parts<Q extends string> = Q extends `${infer U}:${string}` ? U : Q;
+// gets all L2 query snippets for the given L1 part, or `never`
+type L2Parts<
+    Q extends string,
+    P extends string,
+> = Q extends `${P}:${infer U}:${string}` ? U
+    : Q extends `${P}:${infer U}` ? U
+    : never;
 
 // build up all combinations of all L1 fields
-type L1Combinations<Q extends string, L1 extends string> = Combine<
-    L1Fields<Q, L1>,
+type L1Discriminator<Q extends string, L1 extends string> = Combine<
+    L1Fragment<Q, L1>,
```

Comment:
Fields were renamed to fragments, because they make up the fragments of the type to be created. The fact that they are objects which have fields is less speaking.